### DISCUSSION TITLE
Strange behavior of device_{put|get} related to performance

### DISCUSSION THREAD

**Discussion by jjyyxx** [jax-ml/jax#12325] [Q&A]
When I recently worked on a tiny reinforcement library, I encountered strange behavior about wall-time performance when using `device_{put|get}`.

As the context, the most relevant sampling & updating loop (for an off-policy RL algorithm) is as follows. Irrelevant logic (logging, store & replay) are removed or moved out of the loop. **algorithm.update in the loop takes around 1ms on GPU and 3ms on CPU, making any overhead when invoking xla significant.** The `data` tuple contains `(s [256, 20], a [256, 2], d [256], r [256], s' [256, 20])`.
```python
while sample_step < self.total_step:
    # env_timer.enter()
    _, _, done, _ = self.env.step(action)
    # env_timer.exit()

    sample_step += 1
    if done:
        self.env.reset()

    # update_timer.enter()
    self.algorithm.update(update_key, data)
    # update_timer.exit()
```
And `Algorithm.{update|get_action}` where `self.{_update|_get_action}` are two `CompiledFunction`s.
```python
def update(self, key: jax.random.KeyArray, data: Experience) -> Metric:
    # data = jax.device_put(data)
    self.state, info = self._update(key, self.state, data)
    return {k: float(v) for k, v in info.items()}  # {k: float(v) for k, v in jax.device_get(info).items()}

def get_action(self, key: jax.random.KeyArray, obs: np.ndarray) -> np.ndarray:
    action = self._get_action(key, self.state.params.policy, obs)
    return np.asarray(action)  # jax.device_get(action)
```
Environment: Intel(R) Core(TM) i5-8400 CPU @ 2.80GHz; 16GiB RAM; NVIDIA GeForce GTX 1060 3GB; Ubuntu 22.04.1; JAX 0.3.17

I have two questions discovered subsequently and found them not well-documented in JAX documentation. The first question is in fact solved, but I wish a clarification. The second is very strange to me and I have no clue.
1. Similar to `Explicit is better than implicit` in `The Zen of Python`, I used to transfer data between the host and devices explicitly, using `device_{put|get}`. However, through profiling, I discovered that **with the CPU backend**, `device_{put|get}` leads to significant overhead. Simply removing them (in `update` and `get_action`) and substituting them with `asarray` (in `get_action`) leads to around 10% performance improvement. This is counter-intuitive because I was expecting that CPU-only `device_{put|get}` involves at most one copy, which should take only several microseconds.
2. After discovering the first problem, I tested the same micro-optimization **on the GPU backend**, however removing a single line `data = jax.device_put(data)` in `update` led to an unexpected slowdown. More strangely, the performance of `algorithm.update` actually improved a bit but a larger slowdown in the `env.step` function canceled this out. **I have made sure that `env.step` and `algorithm.update` does not have any data dependency** (after simplification). So it's hard to imagine how this could happen. It's worth mentioning that the environment is `PointGoal-v0` from `safety-gym` built upon MuJoCo, and it takes around 1ms per step. With a more lightweight environment like `Pendulum-v1`, such behavior is not observed (the removal leads to pure speed up in this case). See also plots below.

Step time with `PointGoal-v0` (measured with `perf_counter`)
![image](https://user-images.githubusercontent.com/13808089/189642764-cc9c17a8-5038-410d-b277-c27a2f307312.png)

Step time with `Pendulum-v1`
![image](https://user-images.githubusercontent.com/13808089/189642917-7eddf1e2-171a-4fea-9d53-9a57d310b83c.png)

**Edit**: Based on the wavy/rough timing of the "w/o jax.device_put" case, I recently tried disabling Python gc & donate buffer & 64-bytes aligned NumPy array, but the contrast persists. I would like to hear from JAX authors for instructions and suggestions about further investigation.

![image](https://user-images.githubusercontent.com/13808089/190206087-6ef99388-c552-421f-ae64-8ed2f2ab53d7.png)

**Comment by jakevdp** [comment]
I suspect the issue is that when you call `np.asarray` or `__float__` instead of `device_get`, it uses a cached version of the `device_get` operation.

Here's where `__float__` is implemented: https://github.com/google/jax/blob/e5725f1df1831b745efece57f10a017fafaf6951/jax/_src/device_array.py#L296

It calls `self._value` which is a cached property: https://github.com/google/jax/blob/e5725f1df1831b745efece57f10a017fafaf6951/jax/_src/device_array.py#L146-L152

So these methods will only actually transfer data to host a single time, and subsequent calls will be cached.

By contrast, `device_get` directly calls `copy_to_host_async()` on the buffer: https://github.com/google/jax/blob/e5725f1df1831b745efece57f10a017fafaf6951/jax/_src/api.py#L2948

If you look at the source of this, you see that there is no cacheing and it will send the data to the host every time it's called: https://github.com/tensorflow/tensorflow/blob/0845fb8a8ea6227f9963b9800827b22bb44ed3f0/tensorflow/compiler/xla/python/py_buffer.cc#L693-L695

**Reply by jjyyxx** [reply]
Thanks for your reply to my Question 1.

But sorry for the ambiguity that my observation is based on a real running program (a more full-featured training loop than the above snippet mainly targeting Question 2) and profiling rather than a micro-benchmark. (In a real running program, any value transferred twice may indicate unnecessary redundancy and I avoided it for sure.)  At each iteration, the `update` and `get_action` is fed with new onp data, and I don't think there's any room for caching. The main speedup may be attributed to removing explicit `device_put`. The timing of adding/commenting the `device_put` statement:

With `device_put`
```
> time MY_USE_CUDA=0 python scripts/train_rl.py --total_step 5000
warmup took 2605.118 ms
update step 1000
update step 2000
update step 3000
update step 4000
update step 5000

real    0m41.898s
user    1m21.575s
sys     0m5.331s
```

Without `device_put`
```
> time MY_USE_CUDA=0 python scripts/train_rl.py --total_step 5000
warmup took 2589.476 ms
update step 1000
update step 2000
update step 3000
update step 4000
update step 5000

real    0m39.207s
user    1m18.016s
sys     0m4.830s
```
